remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+			return ret;
+	}
+
+	ret = do_get__into_buffer(0, url_component,
+				  type, charset, buffer);
+
+	return ret;
+}
+#endif
+
+/*
+ * Do a single GET (without retry or fallback).
+ */
+static int do_get__into_file(const char *url_base,
+			     const char *url_component,
+			     const struct credential *creds,
+			     struct tempfile *tempfile,
+			     long *http_response_code,
+			     struct strbuf *error_message)
+{
+	struct active_request_slot *slot;
+	struct slot_results results;
+	struct curl_slist *headers = http_copy_default_headers();
+	struct strbuf rest_url = STRBUF_INIT;
+	int ret = HTTP_OK;
+
+	end_url_with_slash(&rest_url, url_base);
+	strbuf_addstr(&rest_url, url_component);
+
+	headers = curl_slist_append(headers,
+				    "X-TFS-FedAuthRedirect: Suppress");
+
+	slot = get_active_slot();
+
+	curl_easy_setopt(slot->curl, CURLOPT_URL, rest_url.buf);
+	curl_easy_setopt(slot->curl, CURLOPT_HTTPHEADER, headers);
+	curl_easy_setopt(slot->curl, CURLOPT_WRITEFUNCTION, fwrite);
+	curl_easy_setopt(slot->curl, CURLOPT_WRITEDATA, (void*)tempfile->fp);
+
+	if (creds && creds->username) {
+		/*
+		 * Force CURL to respect the username/password we provide by
+		 * turning off the AUTH-ANY negotiation stuff.
+		 *
+		 * (AUTH-ANY forces a 401 regardless of what we send).
+		 */
+		curl_easy_setopt(slot->curl, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+		curl_easy_setopt(slot->curl, CURLOPT_USERNAME, creds->username);
+		curl_easy_setopt(slot->curl, CURLOPT_PASSWORD, creds->password);
+	} else {
+		/*
+		 * Turn on the AUTH-ANY negotiation.  This only works with the
+		 * main Git server (because the cache-server doesn't handle 401s).
+		 */
+		curl_easy_setopt(slot->curl, CURLOPT_HTTPAUTH, CURLAUTH_ANY);
+	}
+
+	ret = run_one_slot(slot, &results);
+	curl_easy_getinfo(slot->curl, CURLINFO_RESPONSE_CODE,
+			  http_response_code);
+
+//	trace2_printf("GET: [ret %d] [http %ld] [curl %d '%s'] [res %ld]",
+//		      ret, results.http_connectcode, results.curl_result,
+//		      curl_easy_strerror(results.curl_result),
+//		      *http_response_code);
+
+	strbuf_setlen(error_message, 0);
+	strbuf_addstr(error_message, curl_easy_strerror(results.curl_result));
+
+
+	curl_slist_free_all(headers);
+
+	return ret;
+}
+
+static int do_get__into_file__to_main(const char *url_component,
+				      struct tempfile *tempfile,
+				      struct strbuf *error_message)
+{
+	int ret = HTTP_OK;
+	const struct credential *p_creds = NULL;
+	long http_response_code;
+
+	lookup_main_creds();
+
+	if (rv_data.main_creds.username && *rv_data.main_creds.username)
+		p_creds = &rv_data.main_creds;
+
+	strbuf_setlen(error_message, 0);
+	trace2_region_enter("remote-vfs", "GET/object/main1", NULL);
+	ret = do_get__into_file(rv_data.buf_main_url.buf, url_component,
+				p_creds, tempfile,
+				&http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/main1", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
+
+	if (ret != HTTP_REAUTH)
+		return ret;
```

Comment:
Is `if (ret == HTTP_OK) return HTTP_OK;` really necessary? That case is already handled by the `ret != HTTP_REAUTH` clause, so it just adds distracting code...

Reply:
this aids in debugging, since we can set breakpoints on either the success or failure cases independently.